pydocstringformatter/utils/argument_parsing.py

```diff
@@ -19,24 +20,40 @@ def _parse_command_line_arguments(
 def _register_arguments(version: str) -> argparse.ArgumentParser:
     """Create an argument parser and add all supported arguments"""
     parser = argparse.ArgumentParser(prog="pydocstringformatter")
-
     parser.add_argument("files", nargs="*", type=str)
-
     parser.add_argument(
         "-w",
         "--write",
         action="store_true",
         help="Write the changes to file instead of printing the files to stdout",
     )
-
     parser.add_argument(
         "-v",
         "--version",
         action="version",
         version=version,
         help="Show version number and exit",
     )
+    return parser
+
 
+def _register_arguments_formatters(
+    parser: argparse.ArgumentParser, formatters: List[Formatter]
+) -> argparse.ArgumentParser:
+    """Register a list of formatter, so they can all be deactivated or activated."""
+    for formatter in formatters:
+        help_text = f"ctivate the {formatter.name(delimiter=' ')} formatter"
+        name = formatter.name()
+        parser.add_argument(
+            f"--{name}",
+            action="store_true",
+            dest=name,
+            default=not formatter.optional,
```

Comment:
`store_true` implies a `True` default value. Perhaps we should determine the action based on `formatter.optional`?

Reply:
We have both options ``--closing-quotes`` (that store_true) and  ``--no-closing-quotes`` (that store_false). Whatever the default use using one option or the other will change that default value.

Reply:
Hmm I can see this becoming very complicated.
What if I do:
`pydocstringformatter file --closing-quotes --no-closing-quotes` and `pydocstringformatter file --no-closing-quotes --closing-quotes`?

Reply:
> What if I do: ``pydocstringformatter file --closing-quotes --no-closing-quotes`` and ``pydocstringformatter file --no-closing-quotes --closing-quotes`` ?

Well that would depends on argparse implementation detail. But I think this is an edge case that will never happen if a sentient being is using the program, there's no valid reason to use both. Another way to look at this is that if the parsing args depends on the value of optional, we have a breaking change in our arguments when we change the value of optional. Imo it's better to define both all the time so the API is stable (and the user can always expect both so the args are predictables).

Reply:
It would become a non-edge case if we made `--fix-quotes` which would do `--fix-closing-quotes` and `--fix-opening-quotes` and then have:
`--fix-quotes --no-fix-closing-quotes`.

This is still a bit silly, but I can impinge creating a preset of fixes (possibly `--strict`) of which you might then want to disable one or two. This is a similar issue as in `pylint` where the disabling of messages is dependent on the order of `enable` and `disable`. See https://github.com/PyCQA/pylint/issues/3696. I think that that is actually less intuitive for an end-user.

What about ``--closing-quotes=y`` or ``--closing-quotes=n``? Is possible I would even try to make ``--closing-quotes`` default to ``--closing-quotes=y``.
This does not fully solve the issue of ``--strict`` and ``--closing-quotes=n`` but I'd say we would want an implementation where specific check options suppress any "preset" options.

Reply:
> What about --closing-quotes=y or --closing-quotes=n?

I don't like having to determine if we need an equal and what the truthey and falsey values are for this kind of options. I.e. What do you need to choose between ``--closing-quotes=y`` ``--closing-quotes y``, ``--closing-quotes=1`` or ``--closing-quotes=True``, etc.. So I use ``--x`` and ``--no-x`` which are shown in the --help and are explicit. But It's another solution, it would works too.

Reply:
We can also make ``--closing-quotes`` ``--no-closing-quotes`` options mutually exclusive .

Reply:
Actually the POSIX standard is pretty clear about that, I forgot to make it mutually exclusive. It's been a long time since I worked a fresh new lib 😄

> Guideline 11:
>     The order of different options relative to one another should not matter, unless the options are documented > as mutually-exclusive and such an option is documented to override any incompatible options preceding it. If an > option that has option-arguments is repeated, the option and option-argument combinations should be interpreted in the order specified on the command line.

Reply:
I agree, but having two separate options makes it so much harder to implement and determine what the users wants.
If ``closing-quotes`` default is `True` and due to presets ``closing-quotes`` is `True` and ``no-closing-quotes`` is also `True`, which do we follow?
If ``closing-quotes`` default is `False` and due to presets ``closing-quotes`` is `True` and ``no-closing-quotes`` is also `True`, which do we follow?

It would be much easier if we could have a single option that stored the state of the formatter. What we're looking for is `argparse.BooleanOptionalAction`, but that was only added in `3.9`.

Reply:
> Actually the POSIX standard is pretty clear about that, I forgot to make it mutually exclusive. It's been a long time since I worked on a fresh new lib 😄
> 
> > Guideline 11:
> > The order of different options relative to one another should not matter, unless the options are documented as mutually-exclusive and such an option is documented to override any incompatible options preceding it. If an option that has option-arguments is repeated, the option and option-argument combinations should be interpreted in the order specified on the command line.

We would then need to make ``no-closing-quotes`` exclusive with ``--strict`` which would be defeating the purpose of having presets on which users can then build their specific preferences.

Reply:
Looking at the implementation of `argparse.BooleanOptionalAction` we could actually try and see if we can copy it and use it ourselves on `3.8`? It doesn't look that difficult to maintain and would solve all of our troubles (I hope..)

Reply:
> If closing-quotes default is True and due to presets closing-quotes is True and no-closing-quotes is also True, which do we follow?

Because of ` dest=name,` the two options all store in the same boolean (`closing-quotes`), so in the case you described, the default value is erased and `--no-closing-quote` store False in `closing-quotes`.

Reply:
Ah shit, I missed that. Hmm, then this might indeed be fine. It still feels a bit weird to see `store_true` and then a `default` argument.

Do we know what happens when we call `ArgumentParser.set_defaults`? Will the `store_true` option and its default option or the `store_false` be taken for `dest=name`?

Reply:
> It's been a long time since I worked on a fresh new lib 😄

Same btw, and because of all the issues we're having with configuration parsing in `pylint` (and other libraries) I am bit hesitant about changing anything. I want to be sure we're not missing anything 😄

Reply:
>  I am bit hesitant about changing anything. 

In Pylint I would say the issue is optparse second and having callback first. It's a mess. But I think getting information from the Formatters (Checkers for pylint) is fine. Right now it's just name but we could imagine formatter specific option later. Getting global option from ``_register_arguments`` (somewhere in PyLinter for pylint) is also fine.

Reply:
> .rc file for each directory? The latter would means less duplication.

It would also means more directories if we still want to duplicate because we need to :smile:  tbh I think both are fine

Reply:
Should we do the test refactor in another MR ? Maybe in this MR ?

Reply:
Yeah I think so.

However, we might want to wait on #2 as that would drastically change the output of each tests. With `difflib` which is in `stdlib` it should be fairly trivial to generate `diffs`, but I haven't got round to it (and I'm quite busy this week).
I think because of file names it will be impossible to create `.py.out` with correct diffs, so we might need to move the "functional" tests to checking the written state of a temporary file passed to the test runner?
Lots to do still 😄

Reply:
OK, I'll let you do the diff change, I'm not familiar with this lib. Sorry for being a little too enthusiastic about this fresh project and making you review a lot of code on a busy week 😄